tree_in_context/src/add_context.rs

```diff
@@ -0,0 +1,94 @@
+use crate::{
+    ast::{in_context, infunc},
+    schema::{Assumption, Expr, RcExpr, TreeProgram},
+};
+
+impl TreeProgram {
+    pub fn add_context(&self) -> TreeProgram {
+        TreeProgram {
+            functions: self
+                .functions
+                .iter()
+                .map(|f| f.clone().func_add_context())
+                .collect(),
+            entry: self.entry.clone().func_add_context(),
+        }
+    }
+}
+
+impl Expr {
+    pub(crate) fn func_add_context(self: RcExpr) -> RcExpr {
+        let Expr::Function(name, arg_ty, ret_ty, body) = &self.as_ref() else {
+            panic!("Expected Function, got {:?}", self);
+        };
+        let current_ctx = infunc(name);
+        RcExpr::new(Expr::Function(
+            name.clone(),
+            arg_ty.clone(),
+            ret_ty.clone(),
+            body.add_context(current_ctx),
+        ))
+    }
+
+    fn add_context(self: &RcExpr, current_ctx: Assumption) -> RcExpr {
+        match self.as_ref() {
+            // leaf nodes are constant, empty, and arg
+            Expr::Const(..) | Expr::Empty(..) | Expr::Arg(..) => {
+                in_context(current_ctx, self.clone())
+            }
+            // create new contexts for let, loop, and if
+            Expr::DoWhile(inputs, pred_and_body) => {
+                let new_inputs = inputs.add_context(current_ctx.clone());
+                let new_ctx = Assumption::InLoop(new_inputs.clone(), pred_and_body.clone());
+                RcExpr::new(Expr::DoWhile(
+                    new_inputs,
+                    pred_and_body.add_context(new_ctx),
+                ))
+            }
+            Expr::Let(inputs, body) => {
+                let new_inputs = inputs.add_context(current_ctx.clone());
+                let new_ctx = Assumption::InLet(new_inputs.clone());
+                RcExpr::new(Expr::Let(new_inputs, body.add_context(new_ctx)))
+            }
+            Expr::If(pred, then_case, else_calse) => {
+                let new_pred = pred.add_context(current_ctx.clone());
+                let then_ctx = Assumption::InIf(true, new_pred.clone());
+                let else_ctx = Assumption::InIf(false, new_pred.clone());
+                RcExpr::new(Expr::If(
+                    new_pred,
+                    then_case.add_context(then_ctx),
+                    else_calse.add_context(else_ctx),
+                ))
+            }
+            Expr::Switch(case_num, branches) => {
+                let new_case_num = case_num.add_context(current_ctx.clone());
+                let new_branches = branches
+                    .iter()
+                    .map(|b| b.clone().add_context(current_ctx.clone()))
```

Comment:
nit: do you need this clone? It looks like `add_context` is effectively doing a clone